File: Problem2_RemoveKDigits.java

class Solution {
    public String removeKdigits(String num, int k) {
        
       int len=num.length();
        
        if(k==len)
            return "0";
        
        Stack<Character> stack=new Stack<>();
        
        int i=0;
        
        while(i<num.length())
        {
            //using next greater elememnt pattern in stack
            while(k>0 && !stack.isEmpty() && stack.peek() > num.charAt(i))
            {
                stack.pop();
                k--;
            }
            stack.push(num.charAt(i));
            i++;
        }
        
        // corner case like "1111"
         while(k>0){
            stack.pop();
            k--;            
        }
        
        StringBuilder sb = new StringBuilder();
         while(!stack.isEmpty())
            sb.append(stack.pop());
        
           sb.reverse();
        
         //remove all the 0 at the head
        
        while(sb.length()>1 && sb.charAt(0)=='0')
            sb.deleteCharAt(0);
        
        return sb.toString();
    }
}

File: Problem2_RemoveKDigits.py
# Time Complexity: O(n), where n - length of the
# Space Complexity: O(n), stack

class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []

        # Remove all greater leading digits
        for digit in num:
            while stack and k > 0 and stack[-1] > digit:
                stack.pop()
                k -= 1
            stack.append(digit)

        # If k digits are not yet removed, remove the remaining from stack
        for i in range(k):
            stack.pop()

        # Return the digits from stack as a string and remove the leading zeroes
        result = "".join(stack).lstrip('0')

        if not result:
            return '0'

        return result
